torch/distributions/wishart.py

```diff
@@ -0,0 +1,263 @@
+            new.precision_matrix = self.precision_matrix.expand(cov_shape)
+
+        # Chi2 distribution is needed for Bartlett decomposition sampling
+        new._dist_chi2 = torch.distributions.chi2.Chi2(
+            df=(
+                new.df.unsqueeze(-1)
+                - torch.arange(
+                    self.event_shape[-1],
+                    dtype=new._unbroadcasted_scale_tril.dtype,
+                    device=new._unbroadcasted_scale_tril.device,
+                ).expand(batch_shape + (-1,))
+            )
+        )
+
+        super(Wishart, new).__init__(batch_shape, self.event_shape, validate_args=False)
+        new._validate_args = self._validate_args
+        return new
+
+    @lazy_property
+    def scale_tril(self):
+        return self._unbroadcasted_scale_tril.expand(
+            self._batch_shape + self._event_shape)
+
+    @lazy_property
+    def covariance_matrix(self):
+        return (
+            self._unbroadcasted_scale_tril @ self._unbroadcasted_scale_tril.transpose(-2, -1)
+        ).expand(self._batch_shape + self._event_shape)
+
+    @lazy_property
+    def precision_matrix(self):
+        identity = torch.eye(
+            self._event_shape[-1],
+            device=self._unbroadcasted_scale_tril.device,
+            dtype=self._unbroadcasted_scale_tril.dtype,
+        )
+        return torch.cholesky_solve(
+            identity, self._unbroadcasted_scale_tril
+        ).expand(self._batch_shape + self._event_shape)
+
+    @property
+    def mean(self):
+        return self.df.view(self._batch_shape + (1, 1,)) * self.covariance_matrix
+
+    @property
+    def variance(self):
+        V = self.covariance_matrix  # has shape (batch_shape x event_shape)
+        diag_V = V.diagonal(dim1=-2, dim2=-1)
+        return self.df.view(self._batch_shape + (1, 1,)) * (V.pow(2) + torch.einsum("...i,...j->...ij", diag_V, diag_V))
+
+    def _bartlett_sampling(self, sample_shape=torch.Size()):
+        p = self._event_shape[-1]  # has singleton shape
+
+        # Implemented Sampling using Bartlett decomposition
+        noise = self._dist_chi2.rsample(sample_shape).sqrt().diag_embed(dim1=-2, dim2=-1)
+        i, j = torch.tril_indices(p, p, offset=-1)
+        noise[..., i, j] = torch.randn(
+            torch.Size(sample_shape) + self._batch_shape + (p * (p - 1) // 2,),
+            dtype=noise.dtype,
+            device=noise.device,
+        )
+        chol = self._unbroadcasted_scale_tril @ noise
+        return chol @ chol.transpose(-2, -1)
+
+    def rsample(self, sample_shape=torch.Size(), max_try=10):
+        sample_shape = torch.Size(sample_shape)
+        sample = self._bartlett_sampling(sample_shape)
+
+        # Below part is to improve numerical stability temporally and should be removed in the future
+        is_singular = self.support.check(sample)
+        if self._batch_shape:
+            is_singular = is_singular.amax(self._batch_dims)
+
+        if is_singular.any():
```

Comment:
This will not work as expected under JIT (it will basically use a constant True/False depending on the value during tracing) and will throw a warning. For the JIT version, we will unfortunately need to run this for a fixed number of iterations but maybe we can adjust the default down (this will be wasteful for sure, but I'm not sure if there's a better path):

```
def rsample(..., max_try = None):
    if max_try is None:
        max_try = 3 if torch._C._get_tracing_state() else 10
    if torch._C._get_tracing_state:
        for _ in range(max_try):
        ...
    else:
        ...
```

Reply:
I didn't got understood this part. JIT tests are all passed and I cannot predict potential issue under JIT.
I will apply your suggestion for now.

Reply:
The JIT tracer records the operation on pytorch tensors and outputs an optimized version that gets rid of the python overhead. The issue is that any time we have data dependent control flow, e.g. `if is_singular.any()`, then the tracer will not be able to record the conditional logic. Suppose for the given values, we don't encounter any singular values, then the tracing part will skip the entire for loop and run an optimized version that doesn't run the loop at all. On the other hand, if the initial values are invalid and it needed to run the loop 5 times to get valid values, it will run it for exactly the same number of iterations under JIT. 

I don't see a nice way of handling this, and what I suggested above, just runs the for loop 3 times in all the cases. The JIT tests will not catch these issues, but they will throw a lot of JIT warnings which will give a clue, e.g. for the case above, it complains about "TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!".

Reply:
Thank you for your detailed explanations! It is a comment that many other PyTorch user should see it. I could get a grasp about what the data dependent control flow is thanks to you. It seems far more complex issue arises in JIT. I should study about this topic deeper and analyze why JIT tests did not checked this part. Maybe a another branch for contribution?